### PR TITLE
SHIP-0038: Add shipwright-build-strategy-controller

### PR DESCRIPTION
**Pull request by jkhelil**
Add SHIP to propose the shipwright-build-strategy-controller

### CODE REVIEW

**Review by SaschaSchwarze0 — CHANGES_REQUESTED** [review]
Please rename to 0038. 0037 is https://github.com/shipwright-io/community/pull/136

File: ships/0037-shipwright-build-strategy-controller.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+- Add a new operator to install the Build Strategy catalog
+- Curate or rework the Build Strategies in the upstream build repository when copied to the new catalog repository
+
+## Proposal
+
+We propose to introduce a new CRD in the Shipwright Operator : `ShipwrightBuildStrategy` API.
+
+```yaml
+apiVersion: operator.shipwright.io/v1alpha1
+kind: ShipwrightBuildStrategy
```
Comment (adambkaplan) [inline ships/0037-shipwright-build-strategy-controller.md:61]: Do we really need a new CRD? Can't we extend the existing `ShipwrightBuild` CRD instead?

File: ships/0037-shipwright-build-strategy-controller.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+main.go 
+controllers
+  controller.go
+  add_shipwrightbuild.go
+  add_shipwrightbuildstrategy.go
+  shipwrightbuild
+    shipwrightbuild_controller.go
+  shipwrightbuildstragegy
+     shipwrightbuildstrategy_controller.go
+```
```
Comment (adambkaplan) [inline ships/0037-shipwright-build-strategy-controller.md:125]: This feels like deeper implementation details - we don't want to be overly prescriptive in the proposal when it comes to how this functionality is delivered.

File: ships/0037-shipwright-build-strategy-controller.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+
+### Release Criteria
+
+tbd
+
+### Risks and Mitigations
+
+The default strategy catalog or any url provided catalog can be unreachable if the kubernetes cluster needs
+- a proxy to access to the catalog url
+- it is a disconnected cluster and doesnt have any access to internet
```
Comment (adambkaplan) [inline ships/0037-shipwright-build-strategy-controller.md:178]: Long term, do we need to think about providing OCI artifact resolvers, like Tekton does with task catalogs? Mirroring images is a well established practice for network-constrained environments.

File: ships/0037-shipwright-build-strategy-controller.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  - "@jkhelil"
+reviewers:
+  - "@adambkaplan"
+  - "@qu1queee"
+approvers:
+  - "@adambkaplan"
+  - "@qu1queee"
+creation-date: 2023-10-02
+last-updated: 2023-10-02
+status: implementable
```
Comment (adambkaplan) [inline ships/0037-shipwright-build-strategy-controller.md:19]: @qu1queee for "implementable", should we require that the drawbacks and alternatives sections be filled in? These sections aren't marked as required in the current template.

File: ships/0037-shipwright-build-strategy-controller.md (hunk 1 above)
Comment (jkhelil) [reply ships/0037-shipwright-build-strategy-controller.md:61]: Initially I was thinking about avoiding concurrency issues when updating the CR status, but now that you raise the comment and after thinking about it, We do not need a new CRD, as the controller design is leveraging the optimistic concurrency and if the update fails, we should requeue
I will update the SHIP using the same CRD ShipwrightBuild

File: ships/0037-shipwright-build-strategy-controller.md (hunk 2 above)
Comment (jkhelil) [reply ships/0037-shipwright-build-strategy-controller.md:125]: You are right, I will update the SHIP and add only requireemnt or high level design recommandation
We do want One single manager and multiple controllers assigned  to it, which means e require a new controller for installing the strategies

**Pull request closed without merging**